### PR TITLE
Update cosine score translation for nmslib

### PR DESCRIPTION
**Pull request by VijayanB**
### Description
Nmslib has different score range than exact search. After introducing approximate threshold setting, now, both exact search and approx search can be executed as part of same knn search api. Hence, to keep it consitent, we are overriding the score translation for nmslib for approx search.

### Related Issues
#2319

### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Commits are signed per the DCO using `--signoff`.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by VijayanB** [comment]
Looks like flaky test failure since the same test is successful in Linux and Mac. Create GH issue
https://github.com/opensearch-project/k-NN/issues/2358

File: CHANGELOG.md [hunk 1]
```diff
@@ -28,6 +28,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 * Allow validation for non knn index only after 2.17.0 (#2315)[https://github.com/opensearch-project/k-NN/pull/2315]
 * Release query vector memory after execution (#2346)[https://github.com/opensearch-project/k-NN/pull/2346]
 * Fix shard level rescoring disabled setting flag (#2352)[https://github.com/opensearch-project/k-NN/pull/2352]
+* Use one formula to calculate cosine similarity (#2357)[https://github.com/opensearch-project/k-NN/pull/2357]
```
Comment (jmazanec15) [inline CHANGELOG.md:31]: Probably enhancement instead of bug fix?

File: src/main/java/org/opensearch/knn/index/SpaceType.java [hunk 2]
```diff
@@ -62,7 +62,10 @@ public float scoreToDistanceTranslation(float score) {
     COSINESIMIL("cosinesimil") {
         @Override
         public float scoreTranslation(float rawScore) {
-            return 1 / (1 + rawScore);
+            // To be consistent, we will be using same formula used by lucene as mentioned below
+            // https://github.com/apache/lucene/blob/0494c824e0ac8049b757582f60d085932a890800/lucene/core/src/java/org/apache/lucene/index/VectorSimilarityFunction.java#L73
+            // Here raw score = 1 - cosine(x,y), hence, formula will be updated to below to get final score as similar to Lucene.
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/index/SpaceType.java:67]: I think we need to make this explicit in the contract - maybe add to java doc API. Then, if an engine does something else, it needs to be overridden in the engine specific space translations (see the Faiss specific engine methods for more details).

**Review by jmazanec15 — COMMENTED** [review]
thanks @VijayanB  - can you summarize impact of change in description. For instance, this only impacts nmslib, correct?

File: src/main/java/org/opensearch/knn/index/SpaceType.java (hunk 2 above)
Comment (VijayanB) [reply src/main/java/org/opensearch/knn/index/SpaceType.java]: Updated

File: src/main/java/org/opensearch/knn/plugin/script/KNNScoringSpace.java [hunk 3]
```diff
@@ -144,7 +144,12 @@ public CosineSimilarity(Object query, MappedFieldType fieldType) {
         protected BiFunction<float[], float[], Float> getScoringMethod(final float[] processedQuery) {
             SpaceType.COSINESIMIL.validateVector(processedQuery);
             float qVectorSquaredMagnitude = getVectorMagnitudeSquared(processedQuery);
-            return (float[] q, float[] v) -> 1 + KNNScoringUtil.cosinesimilOptimized(q, v, qVectorSquaredMagnitude);
+            // To be consistent, we will be using same formula used by lucene as mentioned below
+            // https://github.com/apache/lucene/blob/0494c824e0ac8049b757582f60d085932a890800/lucene/core/src/java/org/apache/lucene/index/VectorSimilarityFunction.java#L73
+            return (float[] q, float[] v) -> Math.max(
+                (1.0F + KNNScoringUtil.cosinesimilOptimized(q, v, qVectorSquaredMagnitude)) / 2.0F,
+                0.0F
+            );
```
Comment (navneet1v) [inline src/main/java/org/opensearch/knn/plugin/script/KNNScoringSpace.java:152]: will this create any BWC challenges, mainly in cases when there is a mix cluster. 

Also another thing:
shouldn't this code be 

```suggestion
            // To be consistent, we will be using same formula used by lucene as mentioned below
            // https://github.com/apache/lucene/blob/0494c824e0ac8049b757582f60d085932a890800/lucene/core/src/java/org/apache/lucene/index/VectorSimilarityFunction.java#L73
            return (float[] q, float[] v) -> Math.max(
                (2.0F - KNNScoringUtil.cosinesimilOptimized(q, v, qVectorSquaredMagnitude)) / 2.0F,
                0.0F
            );
```

isn't `KNNScoringUtil.cosinesimilOptimized(q, v, qVectorSquaredMagnitude)` == `rawScore` ?

File: src/test/java/org/opensearch/knn/index/NmslibIT.java [hunk 4]
```diff
@@ -195,6 +195,64 @@ public void testEndToEnd() throws Exception {
...
+        List<KNNResult> knnResults = parseSearchResponse(responseBody, fieldName);
+        assertEquals(k, knnResults.size());
+
+        List<Float> actualScores = parseSearchResponseScore(responseBody, fieldName);
+
+        // both document should have identical score
+        assertEquals(actualScores.get(0), actualScores.get(1), 0.001);
+
+        // Delete index
+        deleteKNNIndex(indexName);
```
Comment (navneet1v) [inline src/test/java/org/opensearch/knn/index/NmslibIT.java:253]: you don't need to delete the index. The base classes IT handles the deletion of indices

File: src/main/java/org/opensearch/knn/plugin/script/KNNScoringSpace.java (hunk 3 above)
Comment (VijayanB) [reply src/main/java/org/opensearch/knn/plugin/script/KNNScoringSpace.java:152]: cosinesimilOptimized  = (dotProduct / (Math.sqrt(normalizedProduct)));

Comment (VijayanB) [reply src/main/java/org/opensearch/knn/plugin/script/KNNScoringSpace.java:152]: > will this create any BWC challenges, mainly in cases when there is a mix cluster.
I don't think it will affect the order of neighbors. I believe we should already have bwc test, let me check and confirm

Comment (navneet1v) [reply src/main/java/org/opensearch/knn/plugin/script/KNNScoringSpace.java:152]: > cosinesimilOptimized = (dotProduct / (Math.sqrt(normalizedProduct)));

Oh wait sorry my bad. This formula is after simplification. Can we please add the docs around how we have simplified it? Just for reference in future.

```math
1 - cos\theta = cosineSimil
```

```math
OS Score = (2 - cosineSimil)/2
```

```math
OSScore = (2 - (1 - cos\theta))/2

```

which simplifies to

```math
OSScore = (1 + cos\theta))/2

```

```math
cos\theta = cosinesimilOptimized
```

Comment (VijayanB) [reply src/main/java/org/opensearch/knn/plugin/script/KNNScoringSpace.java:152]: I see your point since this is not about approx search. Let me add version check.  Good Catch @navneet1v

Comment (VijayanB) [reply src/main/java/org/opensearch/knn/plugin/script/KNNScoringSpace.java]: updated

**Review by navneet1v — APPROVED** [review]
Will we be adding the BWC test for this change after this PR?

**Comment by VijayanB** [comment]
> Will you be adding the BWC test for this change after this PR?

@navneet1v Yes. Will add it as next PR

**Comment by VijayanB** [comment]
Rebased to resolve CHANGE.MD file conflict

**Review by navneet1v — APPROVED** [review]

**Review by jmazanec15 — APPROVED** [review]
